File: LinkedList/go/linkedlist.go
package main

import "fmt"

type Node struct {
	value int
	next  *Node
}

type LinkedList struct {
	head *Node
}

func (l *LinkedList) append(value int) {
	newNode := &Node{value: value, next: nil}

	if l.head == nil {
		l.head = newNode
	} else {
		current := l.head
		for current.next != nil {
			current = current.next
		}
		current.next = newNode
	}
}

func (l *LinkedList) prepend(value int) {
	newNode := &Node{value: value, next: l.head}
	l.head = newNode
}

func (l *LinkedList) insertAfter(prevNode *Node, value int) {
	if prevNode == nil {
		fmt.Println("Previous node cannot be nil.")
		return
	}

	newNode := &Node{value: value, next: prevNode.next}
	prevNode.next = newNode
}

func (l *LinkedList) deleteNode(value int) {
	if l.head == nil {
		return
	}

	if l.head.value == value {
		l.head = l.head.next
		return
	}

	current := l.head
	for current.next != nil {
		if current.next.value == value {
			current.next = current.next.next
			return
		}
		current = current.next
	}
}

func (l *LinkedList) search(value int) bool {
	current := l.head
	for current != nil {
		if current.value == value {
			return true
		}
		current = current.next
	}
	return false
}

func (l *LinkedList) accessByIndex(index int) (int, error) {
	current := l.head
	count := 0
	for current != nil {
		if count == index {
			return current.value, nil
		}
		count++
		current = current.next
	}
	return 0, fmt.Errorf("Index out of range")
}

func (l *LinkedList) size() int {
	current := l.head
	count := 0
	for current != nil {
		count++
		current = current.next
	}
	return count
}

func (l *LinkedList) display() {
	current := l.head
	for current != nil {
		fmt.Printf("%d ", current.value)
		current = current.next
	}
	fmt.Println()
}

func main() {
	linkedList := LinkedList{}
	linkedList.append(1)
	linkedList.append(2)
	linkedList.append(3)
	linkedList.prepend(0)
	linkedList.insertAfter(linkedList.head.next, 1)
	linkedList.display() // Output: 0 1 1 2 3

	linkedList.deleteNode(1)
	linkedList.display() // Output: 0 1 2 3
}

File: LinkedList/java/src/main/java/LinkedList.java
package src.main.java;

public class LinkedList {
    public Node head;

    public static class Node {
        public int value;
        public Node next;

        Node(int value) {
            this.value = value;
            this.next = null;
        }
    }

    // Append a new node with the given value at the end of the linked list
    public void append(int value) {
        Node newNode = new Node(value);
        if (head == null) {
            head = newNode;
        } else {
            Node current = head;
            while (current.next != null) {
                current = current.next;
            }
            current.next = newNode;
        }
    }

    // Prepend a new node with the given value at the beginning of the linked list
    public void prepend(int value) {
        Node newNode = new Node(value);
        newNode.next = head;
        head = newNode;
    }

    // Insert a new node with the given value after the given node
    public void insertAfter(Node prevNode, int value) {
        if (prevNode == null) {
            System.out.println("Previous node cannot be null.");
            return;
        }
        Node newNode = new Node(value);
        newNode.next = prevNode.next;
        prevNode.next = newNode;
    }

    // Delete a node with the given value from the linked list
    public void deleteNode(int value) {
        if (head == null) {
            System.out.println("Linked list is empty.");
            return;
        }
        if (head.value == value) {
            head = head.next;
            return;
        }

        Node current = head;
        while (current.next != null) {
            if (current.next.value == value) {
                current.next = current.next.next;
                return;
            }
            current = current.next;
        }

        System.out.println("Data not found in the linked list.");
    }

    // Search for a node with the given value in the linked list
    public boolean search(int value) {
        Node current = head;
        while (current != null) {
            if (current.value == value) {
                return true;
            }
            current = current.next;
        }
        return false;
    }

    // Access the value of the node at the given index in the linked list
    public int accessByIndex(int index) {
        Node current = head;
        int count = 0;
        while (current != null) {
            if (count == index) {
                return current.value;
            }
            count++;
            current = current.next;
        }
        throw new IndexOutOfBoundsException("Index out of range");
    }

    // Get the number of nodes in the linked list
    public int size() {
        Node current = head;
        int count = 0;
        while (current != null) {
            count++;
            current = current.next;
        }
        return count;
    }

    // Display the contents of the linked list
    public void display() {
        Node current = head;
        while (current != null) {
            System.out.print(current.value + " -> ");
            current = current.next;
        }
        System.out.println("null");
    }

    public static void main(String[] args) {
        LinkedList linkedList = new LinkedList();
        linkedList.append(10);
        linkedList.append(20);
        linkedList.append(30);
        linkedList.prepend(5);
        linkedList.insertAfter(linkedList.head.next, 15);
        linkedList.deleteNode(20);
        linkedList.search(15);
        linkedList.accessByIndex(2);
        linkedList.size();
        linkedList.display();
    }
}
